### DISCUSSION TITLE
Bare metal support

### DISCUSSION THREAD

**Discussion by devorient** [snapcast/snapcast#520] [General]
I would like to create a snapcast client that runs on WiFi microcontrollers, such as the TI CC3200.
I am just starting with it so I was wondering if you have any ideas where to start?
Do you have anything that explains how snapcast internals work without having to dig in the code?

**Comment by Joshfindit** [comment]
Huh. This could be interesting.. Never considered microcontrollers as audio clients... Could it even work on an ESP32 or M0?

**Comment by devorient** [comment]
I think ESP32 should work. An M0 based microcontroller with I2S and DMA should also work.
With the right peripherals audio playback is not CPU heavy. However, memory limitations might be a problem (buffering audio).

**Comment by uSpike** [comment]
Please see related issues: #347 #22

@bridadan created https://github.com/bridadan/libsnapcast

Also python-snapcast.

**Comment by bridadan** [comment]
It is still very very much a work in progress, but I did get audio streaming from an ESP32 this morning!

If you're curious, my ESP-ADF tree is here: https://github.com/bridadan/esp-adf/tree/snapcast
The program is `examples/player/snapcast`.

**Comment by devorient** [comment]
Thanks for the information! Great work @bridadan. I get some hardware, try it out then hopefully we can discuss how I can contribute if you'd like so.

**Comment by bridadan** [comment]
Sounds good! If you have any questions specific to that implementation feel free to raise an issue on my fork there. The next major thing to tackle is timing and synchronization I think. After that its probably down to removing all the hard-coded codec stuff.

**Comment by jorgenkraghjakobsen** [comment]
Hi all  - Have UDP based audio streaming  running over Wifi and PoE/eth on my ESP32s. Very happy to help you getting this ball rolling. I am in the process of writing a Merus Audio driver for the ESP-ADF but effort is on very low pace do to all the other project in the pipe line. Sure your work is a got motivation to take that task up again. 
/j

**Comment by badaix** [comment]
Quoting myself, posted in Issue #22:
> btw: there is a C implementation of Snapcast: [SnapCastC](https://github.com/christf/snapcastc).
I think this implementation uses the same protocol and the player should be compatible to Snapcast.

maybe it helps

**Comment by Gerhardt-Tooling** [comment]
Hello together,
i have been following the same idea idependently. Running Snapcast on ESP32 would be amazing. Thats why I'm designing a Hardware for it. Have a look here:
https://gerhardt-tooling.de/en/projects/

**Comment by Joshfindit** [comment]
> Hello together,
> i have been following the same idea idependently. Running Snapcast on ESP32 would be amazing. Thats why I'm designing a Hardware for it. Have a look here:
> https://gerhardt-tooling.de/en/projects/

Curious to see how this goes, especially if you can get high-quality audio out of it (no noise when audio is not playing, no noise when it’s quiet, clean signal to 25KHz)

**Comment by jorgenkraghjakobsen** [comment]
I have a demo up running - True high quality audio no noise :-)
Demo on WESP32 and MA12070P @12V PVDD. 
We have a combo using ESP32 Wrower and MA12070P @20V in the making
/J   
![IMG_20200306_134547](https://user-images.githubusercontent.com/23186659/76085878-3323f280-5fb3-11ea-98ff-4be47e97cee2.jpg)
![IMG_20200306_134627](https://user-images.githubusercontent.com/23186659/76085655-c577c680-5fb2-11ea-84c7-30226010386b.jpg)

**Comment by Gerhardt-Tooling** [comment]
Hey jorgen, looks good. Did you code it yourself? When my PCBs come i could try to set it up.  Whats the repository for that?

**Comment by bridadan** [comment]
That looks very cool, nice work! Sorry I've been so silent lately, I haven't had much time to work on this. I would like to pick it up again in the future but just not sure when that will be.

**Comment by dmckminard** [comment]
Hello @bridadan ! How can we fast install and test it ? 
I can't make menuconfig, idf.py menuconfig or whatever

**Comment by miLORD1337** [comment]
Hmm, maybe @Devorient could fill in?

**Comment by devorient** [comment]
Hey! I haven't gotten to do any work on this yet. Bought some hardware but never had the time to put it together unfortunately. I would like to start working on it soon but I'm not so sure I'll have the time.

**Comment by jorgenkraghjakobsen** [comment]
I have payed some dedicated night on this and have it up running. I am not coding in the ADF and had to add a ringbuffer between opus decoder and the I2S output driver. 
Multi room sync is not implemented yet - but time stamping and local time sync "just" need to be tied together in a contorl loop. This is def with in reach.

**Comment by TekniskSupport** [comment]
Really nice work! Will definitely try this out ASAP, just have to get my hands on some hardware first.
Again, AWESOME, keep up the good work!

**Comment by jorgenkraghjakobsen** [comment]
https://github.com/jorgenkraghjakobsen/snapclient 
Code is here 
/j

**Comment by TekniskSupport** [comment]
> https://github.com/jorgenkraghjakobsen/snapclient
> Code is here
> /j

Yeah I found it, ordered some hardware and waiting eagerly to try it out

**Comment by jorgenkraghjakobsen** [comment]
@TekniskSupport I have some hardware you can play with if you like - be happy to ship hinsidans :-) 
/j

**Reply by sixtyfive** [reply]
What sort of hardware does one have to acquire to be able to play with your snapclient code, Jorgen?

**Comment by TekniskSupport** [comment]
@jorgenkraghjakobsen Omg what an offer! It will surely go faster than the china order; please hit me up on telegram @organiskPeroxid, or email or whatever and I'll cover any expenses and costs

Nästgårds är ju inte så långt bort
/I

**Comment by jorgenkraghjakobsen** [comment]
Hi All 
1-2 sec sample buffer added - I have a concept to sync audio to the server. 
I will do some code clean up and do a release that demo we have a working ESP32 implementation. 
/Jakobsen

**Comment by AxelJust** [comment]
just a minor mistake in the Instructions

Build

Clone this repo:

git clone https://github.com/jorgenkraghjakobsen/snapcli(**e**)nt 


I'm Curently working on a amplifiler with integrated BMS and USB PD 
At this point I'm Considering to use The ESP32 as Audio Source.
would it be Posible to Ad Bluetooth in adition?

As I am more on the hardware side of things, I orderd a ESP32 Audio Board.
The AI Thinker Audio Kit v2.2 to be exact. It came with a working Bluetooth script for te speakers.
Unfortunatly I can't realy use it now, as None of the example scripts I flasched folowing there instuctions. Smartfone can still be conected to Bluetooth with the changed Name and play music but nothing comes out anywhere.
I estimatet that this is because it dose not use I2S but is meant to be used with internal conversion.
No Idea If they configured it corectly in there code.

Do you think with these aplications there would eaven be enougth power left to do the internal conversion?
I will probably order a Lyrat Board as it uses I2S and comes dyrectly from Espressif.

I would love to hear what you guys think about these points.
Greatings from Erlangen (the Source of MP3)
Axel

**Comment by jorgenkraghjakobsen** [comment]
Hi Alex - For any quality audio application you need external codec to drive an audio amp. 
I am not sure what mistake you are ref to - but yes there at properly quite a few - and i appricated the feedback to nail them. 
Yes you sure can do bluetooth in combination with the snapclient that I am building. I have some code that will be released later. 
I am also very hardware minded and if you like I can help you out on the audio amp side (IFX MA12070P)  - I have some portable designs that I power from LIDL's 20V Li power tools batteries and will rock the party. 
Let me know - Way back i did a lot of programing in Erlang  on Ericsson telecom system - pre internet area :-) 
/Jakobsen

**Comment by AxelJust** [comment]
So cool!
I meant that an e is missing for the github link in the  README.md file.
The amp I am desinging with is the TAS5719 as it has suficient power for a portable speaker and line in + Headphones out.
The next functions i would be curius about would be things like Airplay, Spotify conect,  Mycroft ai or alexa, internet radio ect. ...    but I think I might get in a utopia here.
Do you think it would be posible to ad a ic to ad DAB+ and FM? I could imagine that we could use a usb dongle for that as the ESP32s now also has USB otg.
Greatings
Axel

**Comment by jorgenkraghjakobsen** [comment]
Hi Alex 
Yes the TAS is good for a small portable system - but you will run out of head room if you try to rock the party the that one. 
Airplay 1 is easy just RTP with big buffer, airplay2 is closed source, Spotify connect is not simple but solution in linux should be protable, Mycroft and alexa is available on the esp eco system. 
DAB+ is other chips with i2s output direct to amp or through the ESP. FM is oldschool analog stuff - you need an analog to dig converter for that. 
/j
